### PR TITLE
ssr/have: use opaque constants

### LINKED ISSUE #18017: Performance regression in SSR since SProp support was added to the have tactic
This PR https://github.com/coq/coq/pull/15121 introduced this regression
https://github.com/math-comp/math-comp/pull/1059

Please see https://github.com/math-comp/math-comp/pull/1059#issuecomment-1709829010 for the detailed analysis.
```[tasklist]
### Tasks
- [x] revert to the old code
- [x] duplicate the opaque constant behind have, one per sort
- [x] dispatch to the right constant by checking the sorts of the goal/cut-formula
- [x] when sort-poly lands, coalesce the copies
```

### PR DESCRIPTION
**Pull request by gares**
Fix #18017 
Depends on https://github.com/coq/coq/pull/18596

### CODE REVIEW

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@proux01 @CohenCyril this seems ok now. I did not dare trying to make the constants in Type univpoly.

**Comment by gares** [comment]
@coqbot bench

File: theories/ssr/ssreflect.v
Comment (ppedrot) [inline theories/ssr/ssreflect.v:458]: Isn't this precisely the use case for sort-poly?

Comment (gares) [reply theories/ssr/ssreflect.v:458]: Yes. Is it merged already?

Comment (ppedrot) [reply theories/ssr/ssreflect.v:458]: It's part of 8.19 FWIW.

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
Failures are related to this assertion in kernel/cClosure:
```
let default_evar_handler env = {
  evar_expand = (fun _ -> assert false);
  evar_repack = (fun _ -> assert false);
  evar_irrelevant = (fun _ -> assert false);
  qvar_irrelevant = (fun q ->
      assert (Sorts.QVar.Set.mem q env.env_qualities);
      false);
}
```
@coq/universes-maintainers any idea of what I messed up?

**Comment by ppedrot** [comment]
Which assertion is failing exactly? There is probably some missing state passed to the kernel API, do you have a backtrace?

**Comment by gares** [comment]
Te qvar_irrelevant one

**Comment by gares** [comment]
@coq/ci-maintainers I really don't understand what happens in the bench.
Locally, with my fork of the opam repo, I can install coq-mathcomp-odd-order.dev, but in the bench job it does not work. Similarly for the math-comp (2) packages that are the ones I need to bench.

**Comment by gares** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
I think the problem is that the bench runs ocaml 4.09 but coq-elpi.dev requires >= 4.10 https://github.com/coq/opam/blob/a114c29a0df187a7c6078e0f97607bf6afa6e23a/extra-dev/packages/coq-elpi/coq-elpi.dev/opam#L14
(and non-dev coq-elpi versions have upper bounds on the coq dependency)

File: dev/bench/bench.sh
Comment (SkySkimmer) [inline dev/bench/bench.sh:79]:
```suggestion
: "${skip_coq_tests:=1}"
```
for faster bench iterations while testing

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
Slowdowns are in `Require` so I guess the new code generates more universes or something along these lines.
Any tricks of the trade?

**Comment by SkySkimmer** [comment]
Possibly related to https://github.com/coq/coq/issues/18455

**Comment by gares** [comment]
In my case, the sort-poly constant is only used is non universe polymorphic constants. So I was expecting these extra universe/constraints to be private/hidden/minimized-away. But since it is Require which is slower, I suspect there is some leak out of Qed of this "implementation detail".
I don't know the implementation of sort-poly at all. Is this theory plausible? In the code of the tactic, should I explicitly put a quality/universe instead of generating a  "fresh global" an have Coq do the right thing for me?

**Comment by SkySkimmer** [comment]
Private universes for Qed is only with a univ poly Qed.
If you can put an explicit universe it's better.

**Comment by SkySkimmer** [comment]
You can try doing `Print Universes` after the Qed with and without this PR and see if the size of the output changes.

**Comment by gares** [comment]
I need some help, I quite don't get how to craft the `Instance.t` out of what I have, namely the sort of the argument for `Plemma` and the one for `Pgoal`. I could get the qualities, but I also need levels, and I don't know how to craft `uc` and `ug`.

```ocaml
let qc,qg = ESorts.quality sigma sc, ESorts.quality sigma sg in
let names = UVars.Instance.of_array ([|qc;qg|],[|uc;ug|]) in
let sigma, f = EConstr.fresh_global env sigma ~names f in
```

Here the Coq constant I'm trying to instantiate by hand.

```coq
Polymorphic Lemma ssr_have@{s1 s2|u1 u2|}
  (Plemma : Type@{s1|u1})  (Pgoal : Type@{s2|u2})
  (step : Plemma) (rest : Plemma -> Pgoal) : Pgoal.
```

**Comment by SkySkimmer** [comment]
something like 
~~~ocaml
let sigma, uc = match qc with
  | QConstant (QSProp | QProp) -> sigma, Univ.Level.set
  | _ -> Evd.new_univ_level_variable Evd.univ_flexible sigma
in
(* same for ug using qg *)
~~~
is probably enough.

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot bench
This time without the hack in MC to work around this bug

**Comment by proux01** [comment]
Looks good \o/

**Comment by gares** [comment]
Yup, we can see the gain on solvable, esp in memory, so we were not dreaming ;-)

**Comment by gares** [comment]
@coqbot run full ci

**Review by proux01 — APPROVED** [review]

**Comment by gares** [comment]
@proux01  there is still the bug about the asset false in the kernel.
```
let default_evar_handler env = {
  evar_expand = (fun _ -> assert false);
  evar_repack = (fun _ -> assert false);
  evar_irrelevant = (fun _ -> assert false);
  qvar_irrelevant = (fun q ->
      assert (Sorts.QVar.Set.mem q env.env_qualities); (* this one !!!!!!!!!!!!!!!!! *)
      false);
}
```
@ppedrot

**Comment by SkySkimmer** [comment]
Is there a minimized example?

**Comment by gares** [comment]
@coqbot ci minimize ci-analysis

**Comment by proux01** [comment]
Is the minimizer working, I thought @jashug told us it is currently broken?

**Comment by JasonGross** [comment]
I fixed the minimizer a couple days ago with https://github.com/coq/bot/pull/297

**Comment by JasonGross** [comment]
I'm not sure the minimizer will manage to reduce this anymore unless someone cuts down the runtime to less than 30 seconds by hand.

**Comment by JasonGross** [comment]
> I'm not sure the minimizer will manage to reduce this anymore unless someone cuts down the runtime to less than 30 seconds by hand.

I guess I was wrong, it's making progress

**Comment by proux01** [comment]
@JasonGross is the minimizer doing anything specific with files that are just a bunch of export? I see a lot of
```Coq
Require Import mathcomp.ssreflect.foo.
Require Import mathcomp.ssreflect.bar.
Require Import mathcomp.ssreflect.baz.
Require Import mathcomp.ssreflect.all_ssreflect.
```
where the last file is just exporting all of the above.

**Comment by JasonGross** [comment]
> @JasonGross is the minimizer doing anything specific with files that are just a bunch of export? I see a lot of
> 
> ```coq
> Require Import mathcomp.ssreflect.foo.
> Require Import mathcomp.ssreflect.bar.
> Require Import mathcomp.ssreflect.baz.
> Require Import mathcomp.ssreflect.all_ssreflect.
> ```
> 
> where the last file is just exporting all of the above.

You mean `Require` not `Require Import`, right?  It first recursively adds all transitive `Require` statements, then tries to remove them one by one in dependency order, and then, enter after exhausting all minimization opportunities in the current file, inlines the last non-stdlib require that it hasn't tried yet and tries again.  So it is slowly in lining files.  In the most recent post it is chugging through minimizing the results of inlining `Require mathcomp.algebra.ssrnum.` and the next inline will be `Require mathcomp.algebra.poly` (unless that becomes unused as a result of other minimization).

**Comment by JasonGross** [comment]
Presumably `mathcomp.ssreflect.all_ssreflect` is a transitive dependency via one of the other files being imported, and once the last file that depends on `mathcomp.ssreflect.all_ssreflect` is inlined it'll no longer be necessary to depend on it and the `Require mathcomp.ssreflect.all_ssreflect` will be removed

**Comment by JasonGross** [comment]
It's now inlined everything that depends on `mathcomp.ssreflect.all_ssreflect`, so hopefully in another day or so it'll have removed all the unnecessary ssr dependencies.  The end is coming into sight!

**Comment by SkySkimmer** [comment]
It's done! I'm deleting the previous posts as they make loading this PR quite slow

**Comment by JasonGross** [comment]
Note that the minimizer now contains a direct link to download the bug file from GH Actions, if anyone wants to take a stab at removing the HB.structures dependency.  I've reproduced it here since github actions artifacts eventually expire: [bug.v.zip](https://github.com/coq/coq/files/14101855/bug.v.zip)

**Comment by SkySkimmer** [comment]
The const_relevance isn't properly substituted for opaque constants. https://github.com/coq/coq/issues/18594

**Comment by SkySkimmer** [comment]
Not sure minimizing was any use in the end, the result file is still pretty big and I resorted to printf debugging (I don't know how to get byte files for external plugins to get ocamldebug to work).

**Comment by gares** [comment]
Thanks @SkySkimmer !

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot ci minimize ci-verdi_raft

**Comment by JasonGross** [comment]
Artifacts are missing, try rebasing?

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
I'm looking at CI failures. This one requires some help from @coq/universes-maintainers 
```
Error: Cannot extract nontrivial sort polymorphism
(instantiation of ssreflect.ssr_have_upoly using Prop or SProp).
```
What can I do? What is the problem, actually?

File: theories/ssr/ssreflect.v
Comment (SkySkimmer) [inline theories/ssr/ssreflect.v:459]: This cannot be extracted in a uniform way: when s1 and s2 are Type it should be `fun x f -> f x` but when s1 is Prop or SProp and s2 is Type it should be `fun f -> f __`.
Since we don't know how to do it, we extract as the Type instantiation and reject instantiations with Prop or SProp (unless they're completely erased).

Comment (gares) [reply theories/ssr/ssreflect.v:459]: Hum, then you should extract it twice and call the right one depending on the use, no?

I'm pretty sure Jasmin only uses `have` for proofs, so I'll have to see where the error really comes from.
Would an Extraction directive work (to unblock this PR)?

Comment (SkySkimmer) [reply theories/ssr/ssreflect.v:459]: >Would an Extraction directive work (to unblock this PR)?

I don't think so, the shape of the extracted value changes and I don't think an extraction directive can achieve that.

**Comment by gares** [comment]
The pr was merged on jasmin, but I don't get exactly what I should do @proux01
Do you think I should pin the new master of jasmin? Why is it pinned in the first place?

**Comment by proux01** [comment]
> Do you think I should pin the new master of jasmin?

I'd say yes

> Why is it pinned in the first place?

Because I was hoping for jasmin to get rapidly ported to MC2 and didn't want that to break Coq's CI once more (I already broke it a few times with MC2 release). Unfortunately the port of Jasmin is taking a bit more time than I was expecting.

**Comment by gares** [comment]
@coqbot run full ci

**Comment by JasonGross** [comment]
@coqbot ci minimize

**Comment by JasonGross** [comment]
Oh, ergh, the minimizer is not going to work when the base commit on master and the branch commit have incompatible versions of the CI development.  (Hopefully this is relatively rare?)  Let me make a PR that just bumps jasmin, and then if we rebase this on master after that is merged we can get coqbot to minimize.

**Comment by JasonGross** [comment]
@proux01 @gares Does one of you want to assign https://github.com/coq/coq/pull/19056 and merge [if jasmin passes on CI](https://gitlab.inria.fr/coq/coq/-/jobs/4345148), and then rebase this on top of master once that's merged?

**Comment by gares** [comment]
It does not pass CI, so we broken jasmin *in another PR*.

File: dev/ci/ci-basic-overlay.sh
Comment (gares) [inline dev/ci/ci-basic-overlay.sh:503]:
```suggestion
project jasmin "https://github.com/jasmin-lang/jasmin" "e8380c779b5c284c6d4c654d4ea86c56521a6d4c"
```

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
Metacoq CI failure is related to ssr, but I can't see the link witht this PR.

**Comment by gares** [comment]
@proux01 looks OK to me

**Comment by proux01** [comment]
Great, let's merge then

@coqbot merge now

**Pull request merged**
